src/clj/fluree/db/virtual_graph/index_graph.cljc

```diff
@@ -0,0 +1,97 @@
+            [fluree.db.virtual-graph.proto :as vgproto]
+            [fluree.db.util.json :as json]
+            [fluree.db.util.log :as log]))
+
+#?(:clj (set! *warn-on-reflection* true))
+
+(defn idx-flakes->opts
+  [index-flakes]
+  (reduce
+   (fn [acc idx-flake]
+     (cond
+       (= (flake/p idx-flake) const/$fluree:virtualGraph-name)
+       (assoc acc :vg-name (flake/o idx-flake))
+
+       (and (= (flake/p idx-flake) const/$rdf:type)
+            (not= (flake/o idx-flake) const/$fluree:VirtualGraph))
+       (update acc :type conj (flake/o idx-flake))
+
+       (= (flake/p idx-flake) const/$fluree:query)
+       (try*
+         (assoc acc :query (json/parse (flake/o idx-flake) false))
+         (catch* e
+                 (throw (ex-info (str "Invalid query json provided for Bm25 index, unable to parse: " (flake/o idx-flake))
+                                 {:status 400
+                                  :error  :db/invalid-index}))))
+
+       :else acc))
+   {:type    []
+    :vg-name nil
+    :query   nil}
+   index-flakes))
+
+(defn add-vg-id
+  "Adds the full virtual graph IRI to the index options map"
+  [{:keys [vg-name] :as idx-opts} {:keys [alias] :as _db}]
+  (let [vg-alias (str "##" vg-name)
+        vg-id    (str alias vg-alias)]
+    (assoc idx-opts :id vg-id
+                    :alias vg-alias)))
+
+(defn bm25-idx?
+  [idx-rdf-type]
+  (some #(= % const/$fluree:index-BM25) idx-rdf-type))
+
+(defn create
+  [{:keys [t] :as db} vg-flakes]
+  (let [db-vol         (volatile! db) ;; needed to potentially add new namespace codes based on query IRIs
+        vg-opts        (-> (idx-flakes->opts vg-flakes)
+                           (vg-parse/parse-document-query db-vol)
+                           (add-vg-id db)
+                           (assoc :genesis-t t))
+        {:keys [type alias]} vg-opts
+        db*            @db-vol
+        vg             (cond
+                         ;; add vector index and other types of virtual graphs here
+                         (bm25-idx? type) #?(:clj  (bm25/new-bm25-index db vg-flakes vg-opts)
+                                             :cljs (throw (ex-info "BM25 index not supported in cljs"
+                                                                  {:status 400
+                                                                   :error  :db/invalid-index}))))
+        _              (when (nil? vg)
+                         (throw (ex-info "Unrecognized virtual graph creation attempted."
+                                         {:status 400
+                                          :error  :db/invalid-index})))
+        initialized-vg (vgproto/initialize vg db*)]
+    [db* alias initialized-vg]))
+
+(defn load-virtual-graph
+  [db alias]
+  (or (get-in db [:vg alias])
+      (throw (ex-info (str "Virtual graph requested: " alias " does not exist for the db.")
+                      {:status 400
+                       :error  :db/invalid-query}))))
+
+(defn update-vgs
```

Comment:
This functionality feels like it should eventually be handled by the [`fluree.db.indexer/Indexable` protocol](https://github.com/fluree/db/blob/43013bc5020b3b180330950f02bae404b83f8cdf/src/clj/fluree/db/indexer.cljc#L6-L7). and the [implementation of that protocol for `FlakeDB` instances](https://github.com/fluree/db/blob/43013bc5020b3b180330950f02bae404b83f8cdf/src/clj/fluree/db/flake/flake_db.cljc#L355-L359)

I understand these virtual graphs to be just different kind of indexes sitting alongside the 4 sorted set-based flake indexes, and a `db` is a unified view of all of these indexes. Since the `Indexable` protocol handles updating the sorted set-based indexes after a transaction, I think it could be adapted to also handle updating the bm25 index too, as well as any other indexes that might be defined on a ledger.

Reply:
Agree and I didn't actually try to do that so don't have any reasons as to why that could be difficult or easy.

Couple potential notes that could affect that:
- Unlike our regular indexes which can use novelty to always be current, BM25 is only usable at current 't' once a index is complete, and while fast, it is slower than our other SS-based indexes.
- Mostly because of the above point, you can opt to `sync: false` - which means your query doesn't depend on the latest index, so use the most recently completed one. This means a BM 25 virtual graph needs access to two indexes, and will use one or the other depending on that sync flag (default is `sync: true`).
- Also because of the first point, indexes need to be run with every db update

Reply:
One idea that might accommodate the time difference in updating a vector index vs the instant novelty update is tweaking what we expect to be returned from the `Indexable/-index` method. 

Right now we expect a channel that will eventually contain a single new indexed db instance. Instead, we could expect a channel that could contain multiple indexed dbs. As each indexing process finishes, we return the updated db on the channel, and then we close the channel when all the indexing processes finish.

Downstream code could keep pulling on the channel until it closes and update the latest db by incorporating the new indexes from each value that comes off the channel. 

I think there might be some issues still to hammer out in write heavy scenarios where new transactions come in before the vector index update completes, but I think we can resolve those issues by reusing any previously returned indexing channels that are still open after the `-index` method gets called because a new transaction comes in.